### PR TITLE
Allow Vault variables provider to retry

### PR DESCRIPTION
**Pull request by itowlson**
Maybe it will mitigate that flaky integration test.  Please let it mitigate that flaky integration test.

### CODE REVIEW

File: crates/variables/src/provider/vault.rs [hunk 1]
```diff
@@ -24,12 +25,14 @@ impl VaultProvider {
         token: impl Into<String>,
         mount: impl Into<String>,
         prefix: Option<impl Into<String>>,
+        retry_limit: Option<u8>,
```
Comment (rylev) [inline crates/variables/src/provider/vault.rs:28]: Is there a meaningful semantic difference between `None` and `Some(0)`? Seems like we could just pass a `u8` no?

File: crates/variables/src/provider/vault.rs [hunk 2]
```diff
@@ -52,12 +55,24 @@ impl Provider for VaultProvider {
...
         };
-        match kv2::read::<Secret>(&client, &self.mount, &path).await {
-            Ok(secret) => Ok(Some(secret.value)),
-            // Vault doesn't have this entry so pass along the chain
-            Err(ClientError::APIError { code: 404, .. }) => Ok(None),
-            // Other Vault error so bail rather than looking elsewhere
-            Err(e) => Err(e).context("Failed to check Vault for config"),
+
+        let mut retry_count = 0;
+        loop {
```
Comment (rylev) [inline crates/variables/src/provider/vault.rs:60]: Might be clear to use a `for _ in 0..=self.retry_limit {`

File: crates/variables/src/provider/vault.rs [hunk 3]
```diff
@@ -52,12 +55,24 @@ impl Provider for VaultProvider {
...
+            match kv2::read::<Secret>(&client, &self.mount, &path).await {
+                Ok(secret) => return Ok(Some(secret.value)),
+                // Vault doesn't have this entry so pass along the chain
+                Err(ClientError::APIError { code: 404, .. }) => return Ok(None),
+                // Other Vault error so retry, or if at retry limit bail rather than looking elsewhere
+                Err(e) => {
+                    if retry_count >= self.retry_limit {
```
Comment (rylev) [inline crates/variables/src/provider/vault.rs:67]: I wonder if we should only retry on 5xx errors as other errors are unlikely to be recoverable.

File: crates/trigger/src/runtime_config/variables_provider.rs [hunk 4]
```diff
@@ -70,6 +70,8 @@ pub struct VaultVariablesProviderOpts {
     pub mount: String,
     #[serde(default)]
     pub prefix: Option<String>,
+    #[serde(default)]
+    pub retry_limit: Option<u8>,
```
Comment (rylev) [inline crates/trigger/src/runtime_config/variables_provider.rs:74]: `#[serde(default)]` should allow us to just have this be a `u8`, no?

**Comment by lann** [comment]
> Please let it mitigate that flaky integration test.

It looks like testing-framework supports [Docker `HEALTHCHECK`s](https://docs.docker.com/reference/dockerfile/#healthcheck). I wonder that might solve your flakiness more directly?

**Comment by lann** [comment]
This seems to work:

```
HEALTHCHECK --interval=3s CMD vault status -address=http://127.0.0.1:8200
```

**Comment by lann** [comment]
https://github.com/fermyon/spin/pull/2382

**Comment by itowlson** [comment]
Closing in favour of #2382

**Pull request closed without merging**